BlitzGateway OriginalFile as a Python file-like object

manics opened the pull request:
# What this PR does
Adds a Python read-only file-like object to BlitzGateway OriginalFiles. This allows OriginalFiles to be read straight from OMERO by many Python classes without having to download and save them.

# Testing this PR
1. Upload an OriginalFile
2. Fetch the file in the BlitzGateway
```
f = conn.getObject('OriginalFile', FILEID)
```
3. Pass `f.asFileObj()` to a Python utility that accepts file-like objects. This should be closed after use, and a ContextManager is provided to make this easier. Examples:

## YAML file
```
import yaml
fo = f.asFileObj()
d = yaml.load(fo)
print(d)
fo.close()
```
## Matlab data file `.mat`
```
from scipy.io import loadmat
fo = f.asFileObj()
d = loadmat(fo)
print(d)
fo.close()
```
## CSV file
```
import pandas as pd
with f.asFileObj() as fo:
    csv = pd.read_csv(fo)
    print(csv)
```
## Excel file
```
import pandas as pd
with  f.asFileObj() as fo:
    xl = pd.read_excel(fo)
    print(xl)
```

# Related reading
- https://trello.com/c/N7Xtxk1c/512-file-like-originalfile

manics committed:
Add pythonic file-like class to BlitzGateway

manics committed:
Test BlitzGateway OriginalFileWrapper

manics committed:
Add __iter__ to OriginalFileAsFileObj

This is used by the pandas excel reader

joshmoore commented:
```
./components/tools/OmeroPy/test/integration/gatewaytest/test_wrapper.py:16:1: F401 'Ice' imported but unused

```

mtbc commented on components/tools/OmeroPy/test/integration/gatewaytest/test_wrapper.py:
```diff
@@ -156,6 +158,43 @@ def testDetailsWrapper(self, gatewaywrapper, author_testimg):
+
+        with f.asFileObj() as f2:
+            assert f2.read() == content
+        # Verify closed was automatically called
```
not "close"?

manics committed:
Fix test_wrapper.py typos and travis errors

manics committed:
Move getFileInChunks method into asFileObj

manics committed:
Split testOriginalFileWrapper, test multiple

manics committed:
Fix OriginalFileAsFileObj.close check

For reasons unknown Ice.ObjectNotExistException is being raised instead of omero.ApiUsageException

manics committed:
Fix incorrect BlitzGateway method docs

manics commented:
Updated with some minor factoring so `getFileInChunks` is now handled by `asFileObj`. Also fixed some incorrect documentation on a couple of gateway methods.

manics committed:
Move FileObj into separate test

mtbc commented on components/tools/OmeroPy/test/integration/gatewaytest/test_wrapper.py:
```diff
@@ -156,6 +158,58 @@ def testDetailsWrapper(self, gatewaywrapper, author_testimg):
+        # https://trello.com/c/lC8hFFix/522
+        f1 = self.createTestFile(gatewaywrapper)
+        f2 = self.createTestFile(gatewaywrapper)
+        with f1.asFileObj() as fobj1, f2.asFileObj() as fobj2:
```
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-integration-python/990/testReport/junit/(root)/(empty)/OmeroPy_test_integration_gatewaytest_test_wrapper/ shows a problem: perhaps a Python-2.7-ism?

manics replied:
Thanks, should be fixed now

manics committed:
Use python 2.6 compatible nested with syntax in test

will-moore commented on components/tools/OmeroPy/src/omero/gateway/__init__.py:
```diff
@@ -5230,23 +5279,21 @@ def getFileInChunks(self, buf=2621440):
+        Return a read-only file-like object.
+        Caller must call close() on the file object after use.
+        This can be done automatically by using the object as a
+        ContextManager.
```
Maybe to make this even clearer, provide a tiny example:
```
with  f.asFileObj() as fo:
    for chunk in fo:
        print chunk
```

manics replied:
I've added an example

will-moore commented:
Apart from minor comment, all working well, not leaving statefull services open etc. Good to merge.

manics committed:
Add minimal example to OriginalFileWrapper.asFileObj

rgozim commented:
Looks okay and works as descried.

sbesson commented:
Thanks all. Merging for inclusion in 5.4.9
